source/neuropod/bindings/java/src/main/native/com_uber_neuropod_Neuropod.cc

```diff
@@ -0,0 +1,117 @@
+/* Copyright (c) 2020 UATC, LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "com_uber_neuropod_Neuropod.h"
+
+#include "com_uber_neuropod_LibraryLoader.h"
+#include "jclass_register.h"
+#include "neuropod/core/generic_tensor.hh"
+#include "neuropod/neuropod.hh"
+#include "neuropod/tests/ope_overrides.hh"
+#include "utils.h"
+
+#include <exception>
+#include <memory>
+#include <sstream>
+#include <string>
+
+#include <jni.h>
+
+using namespace neuropod::jni;
+
+JNIEXPORT jlong JNICALL Java_com_uber_neuropod_Neuropod_nativeNew__Ljava_lang_String_2J(JNIEnv *env,
+                                                                                        jclass,
+                                                                                        jstring path,
+                                                                                        jlong   optHandle)
+{
+    try
+    {
+        neuropod::RuntimeOptions opts;
+        if (optHandle != 0)
+        {
+            opts = *reinterpret_cast<neuropod::RuntimeOptions *>(optHandle);
+        }
+        auto                convertedPath = toString(env, path);
+        neuropod::Neuropod *ret           = nullptr;
+        if (isTestMode)
+        {
+            ret = new neuropod::Neuropod(convertedPath, detail::ope_backend_location_overrides, opts);
+        }
+        else
+        {
+            ret = new neuropod::Neuropod(convertedPath, opts);
+        }
+        return reinterpret_cast<jlong>(ret);
+    }
+    catch (const std::exception &e)
+    {
+        throwJavaException(env, e.what());
```

Comment:
I don't follow when it will throw or when this is just a call, like a normal. Can you add a comment in a call where this is not a "throw" ?

Reply:
For example, in the C++ side:

std::vector* a = new std::vector();
delete a;
throwJavaException(env, e.what());
delete a;

This will cause a segfault, because the throwJavaException won't interrupt the C++ side code, both delete is executed.

But for this code:

void jni_foo(env) {
         throwJavaException(JNIEnv *env, "foo");
}

And in the java side:
void main() {
       jni_foo();
}

This program will throw a NeuropodJNIException.

Reply:
Or we can say it throws a Java NeuropodJNIException exception after the JNI call have finished.